foxpuppet/windows/browser/bookmarks/bookmark.py

```diff
@@ -0,0 +1,212 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+"""Contains classes for handling Firefox bookmarks."""
+
+from selenium.webdriver.common.by import By
+from selenium.common.exceptions import NoSuchElementException
+from selenium.webdriver.remote.webelement import WebElement
+from selenium.webdriver.common.keys import Keys
+from foxpuppet.windows.browser.navbar import NavBar
+from typing import TYPE_CHECKING, Optional, TypedDict, List
+
+
+class BookmarkData(TypedDict):
+    """Bookmark properties."""
+
+    name: str
+    url: str
+    tags: Optional[List[str]]
+    keyword: Optional[str]
+
+
+class Bookmark(NavBar):
+    """Handles Bookmark operations in Firefox."""
+
+    if TYPE_CHECKING:
+        from foxpuppet.windows.browser.window import BrowserWindow
+
+    @staticmethod
+    def create(window: "BrowserWindow", root: WebElement) -> "Bookmark":
+        """Create a bookmark object.
+
+        Args:
+            window (:py:class:`BrowserWindow`): Window object this bookmark appears in
+            root (:py:class:`~selenium.webdriver.remote.webelement.WebElement`): WebDriver element object for bookmark
+
+        Returns:
+            :py:class:`Bookmark`: Bookmark instance
+        """
+        with window.selenium.context(window.selenium.CONTEXT_CHROME):
+            return Bookmark(window, root)
+
+    @property
+    def is_bookmarked_star(self) -> bool:
+        """Checks if the current page is bookmarked using the star button.
+
+        Returns:
+            bool: True if the page is bookmarked, False otherwise.
+        """
+        with self.selenium.context(self.selenium.CONTEXT_CHROME):
+            star_button_image = self.selenium.find_element(
+                *BookmarkLocators.STAR_BUTTON_IMAGE
+            )
+            return star_button_image.get_attribute("starred") == "true"
+
+    @property
+    def is_bookmarked_menu(self) -> bool:
+        """Checks if the current page is bookmarked using the main bookmark menu.
+
+        Returns:
+            bool: True if the page is bookmarked, False otherwise.
+        """
```

Comment:
This is saying that if it's bookmarked via the menu it won't also have a star, but will only show in the menu? I don't quite understand this property.

Reply:
Okay, the reason is a result of the way Mozilla URL works. If I paste the Mozilla URL like this `https://www.mozilla.org`, the fully loaded page will be localized, as in, `https://www.mozilla.org/en-GB/`. So if I bookmark the page using the star, the localized URL is what gets bookmarked, not the root URL. However, when I bookmark using the main menu, I use the root URL, which won't match the localized Mozilla URL displayed in the browser, so it won't show as bookmarked If I check using the star button.

Reply:
I wasn't sure if using a localized Mozilla URL might be okay though. Could I just use a localized one?

Reply:
For a test of this logic, a localized URL should be fine.